trace2.c

```diff
@@ -0,0 +1,728 @@
+
+static int tr2main_exit_code;
+
+/*
+ * Our atexit routine should run after everything has finished.
+ *
+ * Note that events generated here might not actually appear if
+ * we are writing to fd 1 or 2 and our atexit routine runs after
+ * the pager's atexit routine (since it closes them to shutdown
+ * the pipes).
+ */
+static void tr2main_atexit_handler(void)
+{
+	struct tr2_tgt *tgt_j;
+	int j;
+	uint64_t us_now;
+	uint64_t us_elapsed_absolute;
+
+	us_now = getnanotime() / 1000;
+	us_elapsed_absolute = tr2tls_absolute_elapsed(us_now);
+
+	/*
+	 * Clear any unbalanced regions so that our atexit message
+	 * does not appear nested.  This improves the appearance of
+	 * the trace output if someone calls die(), for example.
+	 */
+	tr2tls_pop_unwind_self();
+	
+	for_each_wanted_builtin(j, tgt_j) {
+		if (tgt_j->pfn_atexit)
+			tgt_j->pfn_atexit(us_elapsed_absolute,
+					  tr2main_exit_code);
+	}
+
+	tr2_tgt_disable_builtins();
+
+	tr2tls_release();
+	tr2_sid_release();
+	tr2_cfg_free_patterns();
+
+	trace2_enabled = 0;
+}
+
+static void tr2main_signal_handler(int signo)
+{
+	struct tr2_tgt *tgt_j;
+	int j;
+	uint64_t us_now;
+	uint64_t us_elapsed_absolute;
+
+	us_now = getnanotime() / 1000;
+	us_elapsed_absolute = tr2tls_absolute_elapsed(us_now);
+
+	for_each_wanted_builtin(j, tgt_j) {
+		if (tgt_j->pfn_signal)
+			tgt_j->pfn_signal(us_elapsed_absolute, signo);
+	}
+
+	sigchain_pop(signo);
+	raise(signo);
+}
+
+/*****************************************************************
+ * TODO remove this section header
+ *****************************************************************/
+
+void trace2_initialize_fl(const char *file, int line, const char **argv)
+{
+	struct tr2_tgt *tgt_j;
+	int j;
+
+	if (trace2_enabled)
+		return;
+
+	if (!tr2_tgt_want_builtins())
+		return;
+	trace2_enabled = 1;
+
+	tr2_sid_get();
+
+	atexit(tr2main_atexit_handler);
+	sigchain_push(SIGPIPE, tr2main_signal_handler);
+	tr2tls_init();
+
+	/*
+	 * Emit 'version' and 'start' messages on each active builtin target.
+	 */
+	for_each_wanted_builtin(j, tgt_j) {
+		if (tgt_j->pfn_version_fl)
+			tgt_j->pfn_version_fl(file, line);
+		if (tgt_j->pfn_start_fl)
+			tgt_j->pfn_start_fl(file, line, argv);
+	}
+}
+
+int trace2_cmd_exit_fl(const char *file, int line, int code)
+{
+	struct tr2_tgt *tgt_j;
+	int j;
+	uint64_t us_now;
+	uint64_t us_elapsed_absolute;
+
+	code &= 0xff;
+
+	if (!trace2_enabled)
+		return code;
+
+	tr2main_exit_code = code;
+
+	us_now = getnanotime() / 1000;
+	us_elapsed_absolute = tr2tls_absolute_elapsed(us_now);
+
+	for_each_wanted_builtin(j, tgt_j) {
+		if (tgt_j->pfn_exit_fl)
+			tgt_j->pfn_exit_fl(file, line, us_elapsed_absolute, code);
+	}
+
+	return code;
+}
+
+void trace2_cmd_error_va_fl(const char *file, int line,
+			    const char *fmt, va_list ap)
+{
+	struct tr2_tgt *tgt_j;
+	int j;
+
+	if (!trace2_enabled)
+		return;
+
+	for_each_wanted_builtin(j, tgt_j) {
+		if (tgt_j->pfn_error_va_fl)
+			tgt_j->pfn_error_va_fl(file, line, fmt, ap);

```

Comment:
@jeffhostetler  I got a segfault here when running `GIT_TR2_PERF=1 git push origin topic` when a force-push was required.

Reply:
Here is my GDB data:

```
(gdb) run push gh-stolee push-trace
Starting program: /home/stolee/git/git push gh-stolee push-trace
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
14:04:16.895179 common-main.c:36                  | d0 | main                     | version      |     |           |           |            | 2.19.0.gvfs.1.23.g86fac7d809
14:04:16.900169 common-main.c:36                  | d0 | main                     | start        |     |           |           |            |  /home/stolee/git/git push gh-stolee push-trace
14:04:16.900257 repository.c:117                  | d0 | main                     | def_repo     | r1  |           |           |            | worktree:/home/stolee/git
14:04:16.910344 git.c:486                         | d0 | main                     | cmd_verb     |     |           |           |            | push
14:04:17.133183 run-command.c:710                 | d0 | main                     | child_start  |     |  0.238003 |           |            | [ch0] class:? argv: ssh git@github.com 'git-receive-pack '\''derrickstolee/git.git'\'''
14:04:19.032353 run-command.c:956                 | d0 | main                     | child_exit   |     |  2.137168 |  1.899165 |            | [ch0] code:0
To github.com:derrickstolee/git.git
 ! [rejected]              push-trace -> push-trace (non-fast-forward)
error: failed to push some refs to 'git@github.com:derrickstolee/git.git'

Program received signal SIGSEGV, Segmentation fault.
__strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:62
62      ../sysdeps/x86_64/multiarch/strlen-avx2.S: No such file or directory.
(gdb) bt
#0  __strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:62
#1  0x00007ffff73fd4d3 in _IO_vfprintf_internal (s=s@entry=0x7fffffffd6b0, format=format@entry=0x55555575dce0 "failed to push some refs to '%s'", ap=ap@entry=0x7fffffffd830) at vfprintf.c:1643
#2  0x00007ffff74d2169 in ___vsnprintf_chk (s=0x555555a4c080 "failed to push some refs to 'git", maxlen=<optimized out>, flags=flags@entry=1, slen=slen@entry=18446744073709551615,
    format=format@entry=0x55555575dce0 "failed to push some refs to '%s'", args=args@entry=0x7fffffffd830) at vsnprintf_chk.c:63
#3  0x00005555556f86d4 in vsnprintf (__ap=0x7fffffffd830, __fmt=0x55555575dce0 "failed to push some refs to '%s'", __n=<optimized out>, __s=<optimized out>) at /usr/include/x86_64-linux-gnu/bits/stdio2.h:77
#4  strbuf_vaddf (sb=sb@entry=0x7fffffffd870, fmt=0x55555575dce0 "failed to push some refs to '%s'", ap=0x7fffffffd900) at strbuf.c:343
#5  0x0000555555709b1c in fn_error_va_fl (file=0x55555578f9de "usage.c", line=59, fmt=<optimized out>, ap=<optimized out>) at trace2/tr2_tgt_perf.c:226
#6  0x0000555555705566 in trace2_cmd_error_va_fl (file=0x55555578f9de "usage.c", line=59, fmt=0x55555575dce0 "failed to push some refs to '%s'", ap=0x7fffffffd900) at trace2.c:213
#7  0x00005555557202ba in error (err=<optimized out>) at usage.c:213
#8  0x00005555555d25c4 in push_with_options (transport=0x555555a54460, rs=rs@entry=0x5555559f5320 <rs>, flags=flags@entry=0) at builtin/push.c:361
#9  0x00005555555d338c in do_push (push_options=0x5555559e3140 <push_options_config>, flags=0, repo=<optimized out>) at builtin/push.c:426
#10 cmd_push (argc=<optimized out>, argv=<optimized out>, prefix=<optimized out>) at builtin/push.c:636
#11 0x000055555556f2c6 in run_builtin (argv=<optimized out>, argc=<optimized out>, p=<optimized out>) at git.c:490
#12 handle_builtin (argc=<optimized out>, argv=<optimized out>) at git.c:724
#13 0x000055555557032f in run_argv (argv=0x7fffffffe2c0, argcp=0x7fffffffe2cc) at git.c:792
#14 cmd_main (argc=<optimized out>, argv=<optimized out>) at git.c:914
#15 0x000055555556edf2 in main (argc=4, argv=0x7fffffffe528) at common-main.c:46
(gdb) f 6
#6  0x0000555555705566 in trace2_cmd_error_va_fl (file=0x55555578f9de "usage.c", line=59, fmt=0x55555575dce0 "failed to push some refs to '%s'", ap=0x7fffffffd900) at trace2.c:213
213                             tgt_j->pfn_error_va_fl(file, line, fmt, ap);
```

Reply:
I ran `valgrind` which may help a little bit:

```
==33024== Invalid read of size 1
==33024==    at 0x4C32CF2: strlen (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==33024==    by 0x54DD4D2: vfprintf (vfprintf.c:1643)
==33024==    by 0x55B2168: __vsnprintf_chk (vsnprintf_chk.c:63)
==33024==    by 0x2AC6F3: vsnprintf (stdio2.h:77)
==33024==    by 0x2AC6F3: strbuf_vaddf (strbuf.c:343)
==33024==    by 0x2BDB3B: fn_error_va_fl (tr2_tgt_perf.c:226)
==33024==    by 0x2B9585: trace2_cmd_error_va_fl (trace2.c:213)
==33024==    by 0x2D42D9: error (usage.c:213)
==33024==    by 0x1865E1: push_with_options (push.c:362)
==33024==    by 0x1873AB: do_push (push.c:427)
==33024==    by 0x1873AB: cmd_push (push.c:637)
==33024==    by 0x1232C5: run_builtin (git.c:490)
==33024==    by 0x1232C5: handle_builtin (git.c:724)
==33024==    by 0x12432E: run_argv (git.c:792)
==33024==    by 0x12432E: cmd_main (git.c:914)
==33024==    by 0x122DF1: main (common-main.c:46)
==33024==  Address 0x1 is not stack'd, malloc'd or (recently) free'd
==33024==
==33024==
==33024== Process terminating with default action of signal 11 (SIGSEGV)
==33024==  Access not within mapped region at address 0x1
==33024==    at 0x4C32CF2: strlen (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==33024==    by 0x54DD4D2: vfprintf (vfprintf.c:1643)
==33024==    by 0x55B2168: __vsnprintf_chk (vsnprintf_chk.c:63)
==33024==    by 0x2AC6F3: vsnprintf (stdio2.h:77)
==33024==    by 0x2AC6F3: strbuf_vaddf (strbuf.c:343)
==33024==    by 0x2BDB3B: fn_error_va_fl (tr2_tgt_perf.c:226)
==33024==    by 0x2B9585: trace2_cmd_error_va_fl (trace2.c:213)
==33024==    by 0x2D42D9: error (usage.c:213)
==33024==    by 0x1865E1: push_with_options (push.c:362)
==33024==    by 0x1873AB: do_push (push.c:427)
==33024==    by 0x1873AB: cmd_push (push.c:637)
==33024==    by 0x1232C5: run_builtin (git.c:490)
==33024==    by 0x1232C5: handle_builtin (git.c:724)
==33024==    by 0x12432E: run_argv (git.c:792)
==33024==    by 0x12432E: cmd_main (git.c:914)
==33024==    by 0x122DF1: main (common-main.c:46)
```

Reply:
@jeffhostetler The problem is that vsnprintf mutates the va_args as it executes. I verified this by swapping the order of the vreportf and the trace2_cmd_error_va in error_builtin. If I swap them, then your error message appears and the segfault happens during the vreportf.

Reply:
I omitted the va_copy() on the va_list args when I moved code over from my SLOG version.
I just pushed an update to fix this.